### PR TITLE
Bugfix/arsn 252 listing bug versioned bucket

### PR DESCRIPTION
**Pull request by watsaqat**
Addresses https://scality.atlassian.net/browse/S3C-4682.

The `skippingBase` fn is used in Metadata:RepdServer to determine the token from which we can skip to the next key range. Because of [Zenko-1048](https://scality.atlassian.net/browse/ZENKO-1048), this token is not updated correctly and when a key ends with a delimiter, an optimization in RepdServer ends up being allowed to skip all sub-prefixes in the range. This PR adds a version id separator char after tokens ending with the delimiter. Now, when the range is incremented the sub-prefixes  are listed.

### CODE REVIEW

File: lib/algos/list/delimiterMaster.js [hunk 1]
```diff
@@ -95,12 +95,29 @@ class DelimiterMaster extends Delimiter {
...
+                    return FILTER_SKIP;
+                }
+                if (len > 1) {
+                    // key=foo/ or key=foo/bar
+                    if (key[len - 1] === '') {
+                        // key=foo/ : version of a previously seen master key.
+                        return FILTER_SKIP;
+                    }
+                    // Else key=foo/bar (or further nesting as foo/bar/baz/..)
+                    // This case should not be skipped and be handled in this.addContents below.
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: Why so? If the key is `foo/bar` and the value of `this[this.nextContinueMarker]` is `foo/`, we must skip that key as it is part of a prefix already seen, I think?

**Review by jonathan-gramain — COMMENTED** [review]
Can you confirm if this issue only happening with versioned buckets?

Also, make sure you check V1 format, or create a ticket to keep track that we need to do this check.

File: lib/algos/list/delimiterMaster.js (hunk 1 above)
Comment (watsaqat) [reply lib/algos/list/delimiterMaster.js]: this is what gets handled on line 98

Comment (jonathan-gramain) [reply lib/algos/list/delimiterMaster.js]: Can you be more specific

Comment (watsaqat) [reply lib/algos/list/delimiterMaster.js]: I worded this poorly. Should be "should not be skipped *yet*.. ". It does indeed get skipped but this case is handled later in the function. Updated now.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 2]
```diff
@@ -59,6 +59,93 @@ function getListingKey(key, vFormat) {
             assert.strictEqual(delimiter.skipping(), SKIP_NONE);
         });
 
+        it('should not hide keys when a common prefix master is filtered first', () => {
+            // See S3C-4682 for details.
+            // Delimiter will call .filter multiple times with different keys.
+            // It should list them all masters keys except those with delete markers.
+            const delimiter = new DelimiterMaster({
+                prefix: '_Common-Prefix/',
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]:
```suggestion
                prefix: 'Common-Prefix/',
```

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 3]
```diff
@@ -59,6 +59,93 @@ function getListingKey(key, vFormat) {
...
+                value: version.toString(),
+            };
+            // Do not skip master with lexicographically smallest key.
+            assert.strictEqual(delimiter.filter(obj), FILTER_ACCEPT);
+
+            // Skip these with delete markers.
+            // In S3C-4682 there are 514 ids with delete markers and a common prefix.
+            for (let idx = 0; idx < 514; idx++) {
+                // Keys have delete markers and are versioned
+                const keyVersion = `${masterKey}${VID_SEP}${idx}`;
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: Correct me if I'm wrong but I don't think the test is maintaining the key ordering for two reasons, on one hand:

- in https://github.com/scality/Arsenal/pull/1982/files#diff-52fb3163eb0083606af1b564f1535ed6ae456e8b9ee0fbc42be4b22c749ed17dR79 we introduce the `obj.key` ending with the delimiter `/`, and we call `filter` with it
- here we create a version from the `masterKey` which is the same key without the final delimiter, then appending `VID_SEP`, but `VID_SEP` is the null character, lower than `/` in the key ordering. Did you mean to add versions to the same key ending with the delimiter character?

The second reason I see is more practical, we need to pad `idx`, e.g.:

```
const paddedIdx = `0000${idx}`.slice(-4);
```

If I'm not mistaken, the test could have to be reworked to filter keys in order, which maybe could trigger some new testing conditions.

File: lib/algos/list/delimiterMaster.js [hunk 4]
```diff
@@ -95,12 +95,29 @@ class DelimiterMaster extends Delimiter {
...
-                (this.delimiter &&
-                key.startsWith(this[this.nextContinueMarker]))) {
+            if (key === this.prvKey || key === this[this.nextContinueMarker]) {
                 /* master version already filtered */
                 return FILTER_SKIP;
             }
+
+            if (this.delimiter && key.startsWith(this[this.nextContinueMarker])) {
+                // If this[continueMarker] ends with the delimiter then we have to treat it differently.
+                const candidate = key.split(this.delimiter);
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: I don't think splitting the key is really needed for what we want to check, essentially:
- if the key ends with a delimiter character
- if the key contains no delimiter at all in its whole length (not sure why this check)

I'd rather avoid the split call and just do the same kind of checks on the whole key to avoid unnecessary processing and simplify the code.

File: lib/algos/list/delimiterMaster.js [hunk 5]
```diff
@@ -95,12 +95,29 @@ class DelimiterMaster extends Delimiter {
...
+                const len = candidate.length;
+                if (len === 1) {
+                    // case key=foo: this is a version of a previously seen master key.
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: I don't understand the logic, can you try to clarify? Basically I don't see the relationship between not having any delimiter character in the key (hence `len === 1` when we split) and this key being a version of a previously seen master key. I see that a similar case is skipped below when the key ends with a delimiter, but I think if the logic is correct it could deserve some clarification.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 6]
```diff
@@ -59,6 +59,93 @@ function getListingKey(key, vFormat) {
...
+            for (const ob of versionedSuffixes) {
+                const { name, vid } = ob;
+                // Does not contain delete markers but contains versions to simulate turning on versioning.
+                const keyVersion = `${commonPrefix}${name}${VID_SEP}${vid}`;
+                const version = new Version({ versionId: vid });
+                const obj = {
+                    key: getListingKey(keyVersion, vFormat),
+                    value: version.toString(),
+                };
+                const recieved = delimiter.filter(obj);
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: To simplify we could simply check that all results are `FILTER_ACCEPT` since we expect 0 `FILTER_SKIP`

Also a minor spelling typo: ~recieved~ received

Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: In v0 format, all version keys of the same object are preceded by a single master key, so since all those keys are different objects they should all be preceded by a master key (i.e. the same key without `${VID_SEP}${vid}`, and the value of the latest version)

**Review by jonathan-gramain — CHANGES_REQUESTED** [review]

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 6 above)
Comment (watsaqat) [reply tests/unit/algos/list/delimiterMaster.spec.js]: It is harder to debug if you assert FILTER_ACCEPT at each iteration.

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 3 above)
Comment (watsaqat) [reply tests/unit/algos/list/delimiterMaster.spec.js]: if you're just referring to only the keys with isDeleteMarker, we can get rid of those and the behavior wouldn't change. I left them in to more accurately reproduce the bug, but they don't actually matter.

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 2 above)
Comment (watsaqat) [reply tests/unit/algos/list/delimiterMaster.spec.js]: why change this? Its a closer representation to the key that caused the bug than without the underscore. And in general testing with more complicated inputs should lead to us catching more bugs

Comment (jonathan-gramain) [reply tests/unit/algos/list/delimiterMaster.spec.js]: > And in general testing with more complicated inputs should lead to us catching more bugs

I have to disagree, but let's move on it does not really matter just a readability suggestion

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 3 above)
Comment (jonathan-gramain) [reply tests/unit/algos/list/delimiterMaster.spec.js]: It's fine to pass extra keys to more accurately represent the issue, but they should be passed in order, otherwise we're testing an impossible input.

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 6 above)
Comment (jonathan-gramain) [reply tests/unit/algos/list/delimiterMaster.spec.js]: Fair enough, I personally think the readability and simplicity of the test matters more, and if an issue is reproducible it's easy to get down to the cause by adding some debug when the time comes to fix an issue (and it's possible to add context to the `assert.strictEqual` if needed for debugging).

File: lib/algos/list/delimiterMaster.js [hunk 7]
```diff
@@ -122,11 +120,13 @@ class DelimiterMaster extends Delimiter {
              *   when an entry is a PHD avoids the skip on version for the
              *   next entry). In that case we expect the master version to
              *   follow. */
+
             if (key === this.prvPHDKey) {
                 this.prvKey = undefined;
                 return FILTER_ACCEPT;
             }
             this.prvKey = key;
+            this[this.nextContinueMarker] = key;
```
Comment (watsaqat) [inline lib/algos/list/delimiterMaster.js:129]: the key changes are setting the nextContinuationToken to key instead of commonPrefix here and on line 179.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 8]
```diff
@@ -296,7 +514,7 @@ function getListingKey(key, vFormat) {
                 /* When filtered, it should return FILTER_SKIP and set the prvKey. It
                  * should not be added to the result content or common prefixes. */
                 assert.strictEqual(delimiter.filter(obj), FILTER_SKIP);
-                assert.strictEqual(delimiter.NextMarker, undefined);
+                // assert.strictEqual(delimiter.NextMarker, undefined);
```
Comment (watsaqat) [inline tests/unit/algos/list/delimiterMaster.spec.js:517]: I am unsure why having the NextContinuationToken be set to undefined is desirable or needed here and on line 533. Both cases occur after filtering a deleteMarker key.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 9]
```diff
@@ -486,7 +704,7 @@ function getListingKey(key, vFormat) {
                     }),
                     FILTER_ACCEPT);
                 // ...it should return to skipping by prefix as usual
-                assert.strictEqual(delimiter.skipping(), `${inc(DbPrefixes.Replay)}foo/`);
+                assert.strictEqual(delimiter.skipping(), `${inc(DbPrefixes.Replay)}foo/bar${VID_SEP}`);
```
Comment (watsaqat) [inline tests/unit/algos/list/delimiterMaster.spec.js:707]: since we change to returning the key instead of the commonPrefix, this test is changed to reflect this.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 10]
```diff
@@ -46,7 +48,91 @@ function getListingKey(key, vFormat) {
...
+
+function *generateKeys(number) {
+    let idx = 0;
+    let masterKey = '_Common-Prefix/';
+    yield { key: masterKey, isDeleteMarker: false }
+
+    idx++;
+    while (idx < number) {
+        masterKey = `_Common-Prefix/${zpad(idx)}`;
+        yield { key: masterKey, isDeleteMarker: chance.bool() }
```
Comment (miniscruff) [inline tests/unit/algos/list/delimiterMaster.spec.js]: is chance.bool() just a 50/50 random bool? I feel like a simple util function would work instead of a new import but idk.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 11]
```diff
@@ -46,7 +48,91 @@ function getListingKey(key, vFormat) {
...
+    }
+}
+
+/**
+ * Generate raw listing keys in alphabetical order.
+ * @param {Integer} number - how many keys to generate.
+ * @yields {Object} - { key, isDeleteMarker }
+ */
+
+function *generateKeys(number) {
```
Comment (miniscruff) [inline tests/unit/algos/list/delimiterMaster.spec.js]:
```suggestion
function *generateKeys(count) {
```
I would call this var something like count or amount? Just number is kinda vague.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 12]
```diff
@@ -46,7 +48,91 @@ function getListingKey(key, vFormat) {
...
+ * @param {Integer} maxKeys - how many versioned keys to generate
+ * @yields {Object} - { key, isDeleteMarker }
+ */
+function *generateVersionedKeys(masterKey, numKeys) {
+    let [idx, key] = [0, masterKey];
+    while (idx < numKeys) {
+        key = `${masterKey}${VID_SEP}${zpad(idx)}`
+        yield {key, isDeleteMarker: chance.bool()}
+        idx++
+    }
```
Comment (miniscruff) [inline tests/unit/algos/list/delimiterMaster.spec.js]: I would just use a for loop here
```suggestion
    for (let i = 0; i < numKeys; i++) {
        yield {
            key: `${masterKey}${VID_SEP}${zpad(i)}`, 
            isDeleteMarker: chance.bool(),
        }
    }
```
For key you could also do `[masterKey, VID_SEP, zpad(i)].join('')` or just `+`, I never know what is better :smirk:

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 13]
```diff
@@ -46,7 +48,91 @@ function getListingKey(key, vFormat) {
...
+            } else {
+                // stop listing this key range
+                // eslint-disable-next-line no-param-reassign
+                gteparams = params.gte;
+            }
+        }
+    } else {
+        streakLength = 1;
+    }
+    return undefined;
```
Comment (miniscruff) [inline tests/unit/algos/list/delimiterMaster.spec.js]:
```suggestion
```
Unless ESLint complains you don't need this

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 14]
```diff
@@ -59,6 +145,61 @@ function getListingKey(key, vFormat) {
...
+            [...generateKeys(1000)].forEach((ob, idx) => {
+                const {key, isDeleteMarker} = ob;
+
+                // simulate not doing a raw listing when key < params.gte
+                // this represents a key not reaching the read stream from leveldb
+                if (params.gte && key < params.gte) {
+                    throw new Error('Invariant violated. Keys may be missing from listing.');
+                }
+
+                if (key.indexOf(VID_SEP) === -1) { // master key
```
Comment (miniscruff) [inline tests/unit/algos/list/delimiterMaster.spec.js]: Instead of `indexOf` you can do includes: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 15]
```diff
@@ -59,6 +145,61 @@ function getListingKey(key, vFormat) {
...
+                    handleStreak(res, skippingRange, delimiter, idx);
+
+                    let expected = null;
+                    if (vFormat === 'v0') {
+                        expected = isDeleteMarker ? FILTER_SKIP : FILTER_ACCEPT
+                    } else { // v1
+                        expected = FILTER_ACCEPT
+                    }
+                    
+                    assert.equal(res, expected);
```
Comment (miniscruff) [inline tests/unit/algos/list/delimiterMaster.spec.js]: should always use strict assert checks

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 10 above)
Comment (miniscruff) [reply tests/unit/algos/list/delimiterMaster.spec.js]: I see you use chance.integer as well but that is also a one liner :smirk:

**Review by miniscruff — CHANGES_REQUESTED** [review]

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 10 above)
Comment (watsaqat) [reply tests/unit/algos/list/delimiterMaster.spec.js]: fair enough. I'd still rather use a library for these types of tests in general so the homespun versions of chance.bool, chance.integer, etc don't accidentally get written in a way where the distribution is non-uniform. Easy mistake to make, though here it is likely easy enough.

**Review by miniscruff — APPROVED** [review]

**Review by naren-rajendran — APPROVED** [review]

File: lib/algos/list/delimiterMaster.js [hunk 16]
```diff
@@ -131,6 +128,22 @@ class DelimiterMaster extends Delimiter {
         }
 
         this.prvKey = key;
+        if (key === this.prefix) {
+            // key is of the form foo/
+            this.delimiterEndingKeySeen = true;
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: The variable name does not seem to match what it's supposed to mean, as it could refer to any key with an ending delimiter, where it's set only if we've seen a key that `is` the prefix to list. If this is needed, why not calling this variable `keyMatchingPrefixSeen` for example instead.

File: lib/algos/list/delimiterMaster.js [hunk 17]
```diff
@@ -131,6 +128,22 @@ class DelimiterMaster extends Delimiter {
...
+        }
+
+        if (this.delimiterEndingKeySeen === true && // previously seen a key of the form 'prefix/'
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]:
```suggestion
        if (this.delimiterEndingKeySeen && // previously seen a key of the form 'prefix/'
```

File: lib/algos/list/delimiterMaster.js [hunk 18]
```diff
@@ -131,6 +128,22 @@ class DelimiterMaster extends Delimiter {
...
+            !Version.isDeleteMarker(value) && // not delete marker
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: We can remove this check since delete markers are treated earlier in this function and always return.

File: lib/algos/list/delimiterMaster.js [hunk 19]
```diff
@@ -131,6 +128,22 @@ class DelimiterMaster extends Delimiter {
...
+            key.indexOf(VID_SEP) === -1 && // master key
+            key.startsWith(this.prefix) &&
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: same, can be removed as we have an earlier check for keys not starting with the prefix

File: lib/algos/list/delimiterMaster.js [hunk 20]
```diff
@@ -91,13 +92,8 @@ class DelimiterMaster extends Delimiter {
              *   way to know  we should not accept this version. This test is
              *   not redundant with the one at the beginning of this function,
              *   we are comparing here the key without the version suffix,
-             * - key startsWith the previous NextMarker happens because we set
-             *   NextMarker to the common prefix instead of the whole key
-             *   value. (TODO: remove this test once ZENKO-1048 is fixed)
              *   */
-            if (key === this.prvKey || key === this[this.nextContinueMarker] ||
-                (this.delimiter &&
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js:99]: Aren't we introducing a regression by removing this check since ZENKO-1048 is not fixed? If so, can we consider the regression acceptable, or was it causing a more annoying bug in place of what it was supposed to fix? I'd rather try to clarify and keep a comment around related to ZENKO-1048 since there seems to be some corner case here.

File: lib/algos/list/delimiterMaster.js [hunk 21]
```diff
@@ -131,6 +128,22 @@ class DelimiterMaster extends Delimiter {
...
 
         this.prvKey = key;
+        if (key === this.prefix) {
+            // key is of the form foo/
+            this.delimiterEndingKeySeen = true;
+        }
+
+        if (this.delimiterEndingKeySeen === true && // previously seen a key of the form 'prefix/'
+            !Version.isDeleteMarker(value) && // not delete marker
+            key.indexOf(VID_SEP) === -1 && // master key
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]:
```suggestion
            versionIdIndex === -1 && // master key
```

File: lib/algos/list/delimiterMaster.js [hunk 22]
```diff
@@ -131,6 +128,22 @@ class DelimiterMaster extends Delimiter {
...
+            key.startsWith(this.prefix) &&
+            key.indexOf(this.delimiter, this.prefix.length) === -1 // key is of the form 'prefix/foo'
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: I am thinking this check is redundant with the one below (https://github.com/scality/Arsenal/pull/1982/files#diff-9e520dd5b6e739ad1fba611bf9fdffc713fbaf71ed390ad15565a5263578a1f6R151).

If it works I'd rather limit this case specifically to the key equalling the prefix (`key === this.prefix`) which is the case not properly handled and wrongly added as a common prefix, likely the other subkeys would be properly handled below, without needing a specific check. If that's not the case you can keep this but I think it's not needed and potentially could introduce more corner cases, so better try to keep things as simple as possible.

Otherwise if we keep this here I'd recommend re-using the result set in `delimiterIndex` below to avoid a double computation.

File: lib/algos/list/delimiterMaster.js [hunk 23]
```diff
@@ -91,13 +97,9 @@ class DelimiterMaster extends Delimiter {
...
              *   not redundant with the one at the beginning of this function,
              *   we are comparing here the key without the version suffix,
-             * - key startsWith the previous NextMarker happens because we set
-             *   NextMarker to the common prefix instead of the whole key
-             *   value. (TODO: remove this test once ZENKO-1048 is fixed)
              *   */
             if (key === this.prvKey || key === this[this.nextContinueMarker] ||
-                (this.delimiter &&
-                key.startsWith(this[this.nextContinueMarker]))) {
+                this.delimiter && key.startsWith(this[this.nextContinueMarker])) {
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: Why removing the parenthesis?

File: lib/algos/list/delimiterMaster.js [hunk 24]
```diff
@@ -62,6 +63,11 @@ class DelimiterMaster extends Delimiter {
         let key = obj.key;
         const value = obj.value;
 
+        if (key === this.prefix) {
+            // key is of the form prefix/
+            return this.addContents(key, value);
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js:66]: Please also test with a prefix that does not end with a delimiter character, since it should be allowed by the spec. The comment may have to be rephrased or removed then.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 25]
```diff
@@ -59,6 +153,63 @@ function getListingKey(key, vFormat) {
...
+            [...generateKeys(maxKeys)].forEach(ob => {
+                const {key, isDeleteMarker, canSkip} = ob;
+                // simulate not doing a raw listing when key < params.gte
+                // this represents a key not reaching the read stream from leveldb
+                if (state.params.gte && key < state.params.gte) {
+                    return;
+                }
+
+                if (!key.includes(VID_SEP)) { // master key
+                    const version = new Version({ isDeleteMarker });
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: in v1 there is no delete marker key in the master prefix, so we would have to not include them in the listing.

By the way, we should either fix v1 as well or open a ticket for it, if it is not fixed in v1. We can have the test ready for v1 but skip it in the meantime.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 26]
```diff
@@ -46,6 +47,99 @@ function getListingKey(key, vFormat) {
...
+    let masterKey = '_Common-Prefix/';
+    yield { key: masterKey, isDeleteMarker: false, canSkip: false }
+
+    idx++;
+    while (idx < count) {
+        masterKey = `_Common-Prefix/${zpad(idx)}`;
+        const masterKeyWithSubprefix = `_Common-Prefix/${zpad(idx)}/${zpad(idx)}`
+        const baseKey = chance.bool() ? masterKeyWithSubprefix : masterKey;
+        const isDeleteMarker = chance.bool();
+        yield { key: baseKey, isDeleteMarker, canSkip: isDeleteMarker}
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: I am not sure the `canSkip` value is right here: we should be able to skip keys in subprefixes if we have already seen the same subprefix. I think the current code is doing this (in delimiter.js `addCommonPrefix`) but the test does not seem to reflect this.

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 25 above)
Comment (jonathan-gramain) [reply tests/unit/algos/list/delimiterMaster.spec.js]: If there is no such issue with v1, it's fine then with just running the same test on v1.

File: lib/algos/list/delimiterMaster.js (hunk 23 above)
Comment (watsaqat) [reply lib/algos/list/delimiterMaster.js]: copy/paste error. Will update

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 26 above)
Comment (watsaqat) [reply tests/unit/algos/list/delimiterMaster.spec.js]: good catch

File: tests/unit/algos/list/delimiterMaster.spec.js (hunk 25 above)
Comment (watsaqat) [reply tests/unit/algos/list/delimiterMaster.spec.js]: I moved the test to only run under v0; v1 doesn't have this bug and the code change only affects v0 (v1 has separate logic for filtering)

File: lib/algos/list/delimiterMaster.js [hunk 27]
```diff
@@ -122,6 +124,7 @@ class DelimiterMaster extends Delimiter {
              *   when an entry is a PHD avoids the skip on version for the
              *   next entry). In that case we expect the master version to
              *   follow. */
+
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: this extra newline does not seem to be helping, should be removed

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 28]
```diff
@@ -46,6 +48,116 @@ function getListingKey(key, vFormat) {
...
+
+        if (hasSubprefix) {
+            // 'prefix -> liof = -1, pin = 5'
+            // 'prefix/ -> liof = 6, pin = 6'
+            canSkip = state.handleSubprefixKey(baseKey, canSkip)
+        }
+
+        if (state.prefix[state.prefix.length - 1] !== '/') {
+            canSkip = state.handleSubprefixKey(baseKey, canSkip)
+        }
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]:
```suggestion
        if (hasSubprefix || state.prefix[state.prefix.length - 1] !== '/') {
            canSkip = state.handleSubprefixKey(baseKey, canSkip);
        }
```

Also from a style perspective, I'd rather avoid passing `canSkip` to the helper function which should only deal with what it's named for, i.e. subprefix handling, otherwise it becomes hard to follow the logic, we can always use boolean operators to cope with this, something like `canSkip = isDeleteMarker || state.handleSubprefixKey(baseKey)` would likely work.

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 29]
```diff
@@ -488,6 +600,92 @@ function getListingKey(key, vFormat) {
...
             });
+
+            it('should list all subprefixes when handleStreak is applied as in RepdServer', () => {
+                ['_Common-Prefix', '_Common-Prefix/'].forEach(prefix => {
+                    const state = createStreakState(prefix);
+
+                    if (prefix[prefix.length - 1] === '/') {
+                        state.handleSubprefixKey = (baseKey, canSkip) => {
+                            let lastIdx = baseKey.lastIndexOf('/');
+                            // const prefix = baseKey.slice(0, lastIdx);
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: to be removed

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 30]
```diff
@@ -488,6 +600,92 @@ function getListingKey(key, vFormat) {
                 // ...it should return to skipping by prefix as usual
                 assert.strictEqual(delimiter.skipping(), `${inc(DbPrefixes.Replay)}foo/`);
             });
+
+            it('should list all subprefixes when handleStreak is applied as in RepdServer', () => {
+                ['_Common-Prefix', '_Common-Prefix/'].forEach(prefix => {
+                    const state = createStreakState(prefix);
+
+                    if (prefix[prefix.length - 1] === '/') {
+                        state.handleSubprefixKey = (baseKey, canSkip) => {
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: Do we really need two different implementations of this function, why not merging all logic into a single helper function?

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 31]
```diff
@@ -488,6 +600,92 @@ function getListingKey(key, vFormat) {
...
+        
+                            const res = delimiter.filter(obj);
+                            const skippingRange = delimiter.skipping();
+                            handleStreak(res, skippingRange, state);
+        
+                            let expected = null;
+                            if (vFormat === 'v0') {
+                                expected = canSkip ? FILTER_SKIP : FILTER_ACCEPT;
+                            } else { // v1
+                                expected = FILTER_ACCEPT
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: I don't understand why in v1 all master keys are accepted. I would think that the behavior should be similar for v0 and v1 when it comes to master keys handling.

File: lib/algos/list/delimiterMaster.js [hunk 32]
```diff
@@ -33,6 +33,7 @@ class DelimiterMaster extends Delimiter {
         this.prvKey = undefined;
         this.prvPHDKey = undefined;
         this.inReplayPrefix = false;
+        this.delimiterEndingKeySeen = false;
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: looks like this should be removed

File: lib/algos/list/delimiterMaster.js [hunk 33]
```diff
@@ -91,13 +97,9 @@ class DelimiterMaster extends Delimiter {
              *   way to know  we should not accept this version. This test is
              *   not redundant with the one at the beginning of this function,
              *   we are comparing here the key without the version suffix,
-             * - key startsWith the previous NextMarker happens because we set
```
Comment (jonathan-gramain) [inline lib/algos/list/delimiterMaster.js]: Why removing this comment

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 34]
```diff
@@ -46,6 +47,111 @@ function getListingKey(key, vFormat) {
     return assert.fail(`bad vFormat ${vFormat}`);
 }
 
+const MAX_STREAK_LENGTH = 100
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: I though that the linter would complain on this, maybe the rules have been updated? Anyway I think it's a good practice to end all instructions with a semicolon.

```suggestion
const MAX_STREAK_LENGTH = 100;
```

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 35]
```diff
@@ -46,6 +47,111 @@ function getListingKey(key, vFormat) {
...
+const createStreakState = (prefix, vFormat) => {
+    return {
+        vFormat,
+        prefix,
+        prefixes: new Set(),
+        params: {},
+        streakLength: 0,
+        gteparams: null,
+    }
+}
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js:60]:
```suggestion
};
```

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 36]
```diff
@@ -46,6 +47,111 @@ function getListingKey(key, vFormat) {
...
+        const isDeleteMarker = state.vFormat === 'v0' && chance.bool() ;
+        const baseKey = hasSubprefix ? masterKeyWithSubprefix : masterKey;
+
+        let canSkip = isDeleteMarker;
+        if (hasSubprefix || state.prefix[state.prefix.length - 1] !== '/') {
+            canSkip =  canSkip || state.handleSubprefixKey(baseKey);
+        }
+
+        yield { key: baseKey, isDeleteMarker, canSkip }
+        idx++
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]:
```suggestion
        yield { key: baseKey, isDeleteMarker, canSkip };
        idx++;
```

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 37]
```diff
@@ -46,6 +47,111 @@ function getListingKey(key, vFormat) {
...
+
+        const versioned = chance.integer({ min: 0, max: 200});
+        const allowedVersionedKeys = Math.min(count - idx, versioned)
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]:
```suggestion
        const allowedVersionedKeys = Math.min(count - idx, versioned);
```

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 38]
```diff
@@ -59,6 +165,71 @@ function getListingKey(key, vFormat) {
...
+            ['_Common-Prefix', '_Common-Prefix/'].forEach(prefix => {
+                const state = createStreakState(prefix, vFormat);
+                state.handleSubprefixKey = baseKey => {
+                    const isLastDelim = state.prefix[state.prefix.length - 1] === '/';
+                    let lastIdx = isLastDelim ? baseKey.lastIndexOf('/') : state.prefix.length;
+                    const prefix = isLastDelim ? getCommonPrefix(baseKey, '/', lastIdx): baseKey.slice(0, lastIdx);
+                    if (state.prefixes.has(prefix) || (!isLastDelim && baseKey.length > lastIdx)) {
+                        return true;
+                    } else {
+                        state.prefixes.add(prefix)
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]:
```suggestion
                        state.prefixes.add(prefix);
```

File: tests/unit/algos/list/delimiterMaster.spec.js [hunk 39]
```diff
@@ -59,6 +165,71 @@ function getListingKey(key, vFormat) {
...
+        it('should list all subprefixes when handleStreak is applied as in RepdServer', () => {
+            ['_Common-Prefix', '_Common-Prefix/'].forEach(prefix => {
+                const state = createStreakState(prefix, vFormat);
+                state.handleSubprefixKey = baseKey => {
+                    const isLastDelim = state.prefix[state.prefix.length - 1] === '/';
+                    let lastIdx = isLastDelim ? baseKey.lastIndexOf('/') : state.prefix.length;
+                    const prefix = isLastDelim ? getCommonPrefix(baseKey, '/', lastIdx): baseKey.slice(0, lastIdx);
+                    if (state.prefixes.has(prefix) || (!isLastDelim && baseKey.length > lastIdx)) {
+                        return true;
+                    } else {
```
Comment (jonathan-gramain) [inline tests/unit/algos/list/delimiterMaster.spec.js]: The `else` is not needed after the `return`

**Review by jonathan-gramain — COMMENTED** [review]
There's something strange with the linter rules, anyway let's fix this and get this merged

**Review by jonathan-gramain — COMMENTED** [review]
ok with the code, please squash commits

**Review by jonathan-gramain — COMMENTED** [review]
Can you please add a description of the bugfix in the commit message (you can use the PR's description for example)

**Review by jonathan-gramain — APPROVED** [review]
